### PR TITLE
Using pycaret datetime features

### PR DESCRIPTION
**Pull request by joker2411**
## Description of the change

> Ticket [Link](https://linear.app/rudderstack/issue/PRML-870/use-pycaret-datetime-features).

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/predictions/profiles_mlcorelib/connectors/RedshiftConnector.py [hunk 1]
```diff
@@ -33,7 +33,11 @@ def __init__(self, creds: dict, folder_path: str) -> None:
             "arraytype": {"super": None},
             "booleantype": {"boolean": None, "bool": None},
         }
-        self.dtype_utils_mapping = {"numeric": "float", "categorical": "str"}
+        self.dtype_utils_mapping = {
+            "numeric": "float",
+            "categorical": "str",
+            "timestamp": "timestamp without time zone",
```
Comment (dpatchigolla) [inline src/predictions/profiles_mlcorelib/connectors/RedshiftConnector.py:39]: Does this cover different types of timestamp datatypes? with time zone, if it is not explicitly mentioned, or datetype etc? How are they handled?

File: src/predictions/profiles_mlcorelib/connectors/SnowflakeConnector.py [hunk 2]
```diff
@@ -57,6 +57,7 @@ def __init__(self, creds: dict) -> None:
         self.dtype_utils_mapping = {
             "numeric": "DecimalType",
             "categorical": "StringType",
+            "timestamp": "TimestampType",
```
Comment (dpatchigolla) [inline src/predictions/profiles_mlcorelib/connectors/SnowflakeConnector.py:60]: Same as for redshift - what happens to datetype, timestamptype with tz, without tz etc? There are many datatypes in snowpark associated with datetime/time
https://docs.snowflake.com/ko/developer-guide/snowpark/reference/python/latest/types

Timesatmp, TimestampType, TimestampTimeZone, DateType etc.

File: src/predictions/profiles_mlcorelib/ml_core/preprocess_and_predict.py [hunk 3]
```diff
@@ -130,6 +126,11 @@ def load_model(filename: str):
         return model
 
     def predict_helper(df, pkl_model_file_name: str) -> Any:
+        df.columns = features
```
Comment (dpatchigolla) [inline src/predictions/profiles_mlcorelib/ml_core/preprocess_and_predict.py:129]: Why is this moved from prediction_scores class to here? And I think  the `astype("float64")` was added to prevent some issue in bigquery datatypes where it had nulls of pd.<NA> type. This might cause an issue dropping this line.

File: src/predictions/profiles_mlcorelib/ml_core/preprocess_and_train.py [hunk 4]
```diff
@@ -116,10 +113,6 @@ def prepare_feature_table(
         )
         feature_table = connector.drop_cols(feature_table, [trainer.label_column])
 
-        for col in timestamp_columns:
```
Comment (dpatchigolla) [inline src/predictions/profiles_mlcorelib/ml_core/preprocess_and_train.py:119]: Do we still have the datediff feature, or are we dropping that?

File: src/predictions/profiles_mlcorelib/utils/utils.py [hunk 5]
```diff
@@ -160,7 +160,7 @@ def get_feature_table_column_types(
         input_column_types["numeric"][col] = dtype_mapping["numeric"]
 
     for col in input_column_types["timestamp"]:
-        input_column_types["numeric"][col] = dtype_mapping["numeric"]
+        input_column_types["timestamp"][col] = dtype_mapping["timestamp"]
```
Comment (dpatchigolla) [inline src/predictions/profiles_mlcorelib/utils/utils.py:163]: do we drop the datediff numeric features? if not, don't we need to have that numeric type casting too?

**Review by dpatchigolla — COMMENTED** [review]
Some questions that need clarification - 
also, do we have the datetime and timestamp features in all three warehouses?

File: src/predictions/profiles_mlcorelib/ml_core/preprocess_and_predict.py (hunk 3 above)
Comment (joker2411) [reply src/predictions/profiles_mlcorelib/ml_core/preprocess_and_predict.py:129]: It was being duplicated in both the `predict_scores` and `predict_scores_rs` functions, so I just moved it to the child function.

File: src/predictions/profiles_mlcorelib/utils/utils.py (hunk 5 above)
Comment (joker2411) [reply src/predictions/profiles_mlcorelib/utils/utils.py:163]: we didn't drop those. Instead I removed that piece of code and now we are directly passing timestamp features to train fn in pycaret. It's processing them inside the fn.

File: src/predictions/profiles_mlcorelib/ml_core/preprocess_and_train.py (hunk 4 above)
Comment (joker2411) [reply src/predictions/profiles_mlcorelib/ml_core/preprocess_and_train.py:119]: no, we are not converting them now.

**Review by dpatchigolla — APPROVED** [review]
